core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -368,35 +407,44 @@ public void onFailure(Throwable e) {
 
         @Override
         protected void doRun() {
-            if (checkBlocks() == false) {
+            // shardID added to the request, as a side effect of calling resolveRequest
+            if (checkBlocksAndResolveRequest() == false) {
```

Comment:
This is wonderful! Now `observer#observedState()` is invoked in exactly one place in the source file.